waiter/docs/service-references.md

```diff
@@ -0,0 +1,29 @@
+# Service References
+
+To process a http request, Waiter needs to resolve a request to a service description.
+This service description tells Waiter how to locate the service instance to use to process the current request.
+
+Waiter uses request headers to construct the service description using the `ServiceDescriptionBuilder`.
+It is possible for the service to be reachable from various combinations of headers:
+  1. token - single or multiple
+  1. on-the-fly services without tokens
+  1. on-the-fly services that extend tokens, e.g. run-as-requester
+
+## Service GC
+
+By default, Waiter relies on the service idle timeout to GC services after periods of inactivity (not receiving requests).
+However, service reachable only via references, e.g. tokens, can be GC-ed eagerly if the reference has been updated.
+
+When a service description is constructed from a request, the service references are also updated.
+These references are available as the `:references` key in the descriptor and are annotated by a `:type` parameter.
+The service GC process checks these references by type and marks a service as a candidate for eager GC
+  if _all_ references used to access the service are stale.
```

Comment:
Running service two references: token and image.  Image does stale.  New service goes up for new image.  Are we not GC'ing the old service, since it's token is not stale?

Reply:
In the proposed example: the token reference entry will have additional metadata that it introduced an image and the token staleness check function will have to account for it.